tests/extmod/machine_uart_irq_txidle.py

```diff
@@ -0,0 +1,82 @@
+# Test machine.UART.IRQ_TXIDLE firing after a transmission.
+# Does not require any external connections.
+
+try:
+    from machine import UART
+
+    UART.IRQ_TXIDLE
+except (ImportError, AttributeError):
+    print("SKIP")
+    raise SystemExit
+
+import time, sys
+
+# Configure pins based on the target.
+hard_irq = False
```

Comment:
Maybe add a comment: "use hard IRQ on ports that need to respond immediately to catch the TXIDLE flag"

Reply:
OK.

Reply:
Ideally we could remove these kinds of small inconsistencies between ports and make the IRQ behave as similar as possible across the ports.

Just to be clear, the hard IRQ is needed only on rp2 when TX/RX have a connection between them (which isn't required for this test but is for the other ones)?

And for samd?

Maybe would be good to make the comment here more detailed, with the above details, so someone who wants to eventually fix things can know exactly what to look for, what tests to run to check if hard IRQ is really needed.

Reply:
> Just to be clear, the hard IRQ is needed only on rp2 when TX/RX have a connection between them (which isn't required for this test but is for the other ones)?

Yes. The NRF port has similar issues, but does not support hard irq. The SAMD port is fine.

Reply:
I assume that the IRQ_TXIDLE is triggered while still data is sent. For RP2 this is doumented as 4-5 bytes, but for SAMD21 this is true for one byte as well. Then, when the IRQ callback is called, the IRQ_TXIDLE flag of self->irq_flags is overwritten by the byte TX interrupt.

Reply:
> we could change the NRF port to accept & ignore the hard irq argument instead raising an error.

All ports should (eventually) be using the shared code in `shared/runtime/mpirq.c`, and then they would all support the `hard` argument (I guess).  But getting all this consistent now may be too big of a job and we'd never finish :)

> Then, when the IRQ callback is called, the IRQ_TXIDLE flag of self->irq_flags is overwritten by the byte TX interrupt.

So maybe there's something to fix here: if the user doesn't want to respond to TX IRQ's (there's no such thing as far as the user is concerned) then `self->irq_flags` should not be updated when this IRQ occurs.

Reply:
> All ports should (eventually) be using the shared code in shared/runtime/mpirq.c, and then they would all support the hard argument (I guess). 

The NRF port uses the code from  shared/runtime/mpirq.c but raises an error if the `hard` argument is used.

> if the user doesn't want to respond to TX IRQ's (there's no such thing as far as the user is concerned) then self->irq_flags should not be updated when this IRQ occurs.

With the RX/TX loopback it's the RX event which is triggered. And the uart.irq() callback and the uart IRQ handler are not synchronized. Ideally the respective flag for an event would only be cleared if the callback has finished. But that is not possible for a scheduled callback.

Reply:
I reworked the handling of the the flags reported by irq.flags(). It will only be changed when a event is present which is expected by the trigger. With that change, hard=True is not needed for the TXIDLE test, and the RX/TX loopback does not cause any problem.

Reply:
Excellent!